pkg/actor/resize_pvc.go

```diff
@@ -0,0 +1,280 @@
+// replaced with shorter times to allow testing of Wait___ functions without
+// waiting the entire default period
+var backoffFactory = defaultBackoffFactory
+
+func defaultBackoffFactory(maxTime time.Duration) backoff.BackOff {
+	b := backoff.NewExponentialBackOff()
+	b.MaxElapsedTime = maxTime
+	return b
+}
+
+// newResizePVC creates and returns a new resizePVC struct
+func newResizePVC(scheme *runtime.Scheme, cl client.Client, config *rest.Config) Actor {
+	return &resizePVC{
+		action: newAction("resize_pvc", scheme, cl),
+		config: config,
+	}
+}
+
+// resizePVC resizes a PVC
+type resizePVC struct {
+	action
+
+	config *rest.Config
+}
+
+// Handles returns true if the DB is initialized
+func (rp *resizePVC) Handles(conds []api.ClusterCondition) bool {
+	return condition.False(api.NotInitializedCondition, conds) && utilfeature.DefaultMutableFeatureGate.Enabled(features.ResizePVC)
+}
+
+// Act in this implementation resizes PVC volumes of a CR sts.
+func (rp *resizePVC) Act(ctx context.Context, cluster *resource.Cluster) error {
+	log := rp.log.WithValues("CrdbCluster", cluster.ObjectKey())
+
+	// If we do not have a volume claim we do not have PVCs
+	if cluster.Spec().DataStore.VolumeClaim == nil {
+		log.Info("Skipping PVC resize as VolumeClaim does not exist")
+		return nil
+	}
+
+	// Get the sts and compare the sts size to the size in the CR
+	key := kubetypes.NamespacedName{
+		Namespace: cluster.Namespace(),
+		Name:      cluster.StatefulSetName(),
+	}
+	statefulSet := &appsv1.StatefulSet{}
+	if err := rp.client.Get(ctx, key, statefulSet); err != nil {
+		return errors.Wrap(err, "failed to fetch statefulset")
+	}
+
+	// TODO statefulSetIsUpdating is not quite working as expected.
+	// I had to check status.  We should look at the update code in partition update to address this
+	if statefulSetIsUpdating(statefulSet) {
+		return NotReadyErr{Err: errors.New("resize statefulset is updating, waiting for the update to finish")}
+	}
+
+	status := &statefulSet.Status
+	if status.CurrentReplicas == 0 || status.CurrentReplicas < status.Replicas {
+		log.Info(" resize pvc statefulset does not have all replicas up")
+		return NotReadyErr{Err: errors.New("resize pvc statefulset does not have all replicas up")}
+	}
+
+	// Maybe this should be an error since we should not have this, but I wanted to check anyways
+	if len(statefulSet.Spec.VolumeClaimTemplates) == 0 {
+		log.Info("Skipping PVC resize as PVCs do not exist")
+		return nil
+	}
+
+	clientset, err := kubernetes.NewForConfig(rp.config)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create kubernetes clientset")
+	}
+
+	stsStorageSizeDeployed := statefulSet.Spec.VolumeClaimTemplates[0].Spec.Resources.Requests.Storage()
+	stsStorageSizeSet := cluster.Spec().DataStore.VolumeClaim.PersistentVolumeClaimSpec.Resources.Requests.Storage()
+
+	// If the sizes match do not resize
+	if stsStorageSizeDeployed.Equal(stsStorageSizeSet.DeepCopy()) {
```

Comment:
You lost me.  I am not certain what you mean.

Reply:
@chrisseto ping

Reply:
In a naive read through of this code, this cases jumps out to me. It looks like Some PVCs might not get resized if something goes wrong.
If the order was 1. Update STS, 2. Update PVCs; this check could result in some PVCs not being updated.

To ease some future confusion, it may be helpful to update this comment point out that updating the STS update is performed last?